File: griffon-app/views/ru/dkolmogortsev/ControlPanelView.java
package ru.dkolmogortsev;

import griffon.core.artifact.GriffonView;
import griffon.inject.MVCMember;
import griffon.metadata.ArtifactProviderFor;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleListProperty;
import javafx.fxml.FXML;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.geometry.Side;
import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.CustomMenuItem;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.layout.ColumnConstraints;
import javafx.scene.layout.GridPane;
import org.codehaus.griffon.runtime.javafx.artifact.AbstractJavaFXGriffonView;
import org.reactfx.EventStream;
import org.reactfx.EventStreams;
import org.reactfx.Subscription;
import ru.dkolmogortsev.task.search.SearchFields;

@ArtifactProviderFor(GriffonView.class)
public class ControlPanelView extends AbstractJavaFXGriffonView
{
    @FXML
    Button startButton;
    @FXML
    Label timerLabel;
    @FXML
    TextField taskDescription;
    @FXML
    TextField taskName;
    @MVCMember
    private ControlPanelController controller;
    @MVCMember
    private ControlPanelModel model;
    @MVCMember
    private ControlAndTaskView parentView;
    private BooleanProperty started = new SimpleBooleanProperty(false);
    private Subscription descInputSubscription;
    private SimpleListProperty<String> tasksToShow = new SimpleListProperty<>();

    private ContextMenu menu = new ContextMenu();
    private GridPane customPopup = new GridPane();

    @Override
    public void initUI()
    {
        Node node = init();
        timerLabel.textProperty().bind(model.getTimerTextProp());
        started.bind(model.taskStartedProperty());
        tasksToShow.bind(model.tasksProperty());

        taskDescription.textProperty().bindBidirectional(model.taskDescriptionProperty());
        taskName.textProperty().bindBidirectional(model.taskNameProperty());

        toStartButton();
        parentView.getAnchorPane().addRow(0, node);
        started.addListener(((observable, oldValue, newValue) ->
        {
            if (newValue)
            {
                toStopButton();
                return;
            }
            toStartButton();
        }));

        EventStream<String> inputStream = EventStreams.valuesOf(taskDescription.textProperty());

        EventStreams.valuesOf(taskDescription.focusedProperty()).subscribe(aBoolean ->
        {
            if (aBoolean)
            {
                descInputSubscription = inputStream
                        .subscribe((searchStr) -> controller.search(searchStr, SearchFields.DESCRIPTION));
                return;
            }
            if (descInputSubscription != null)
            {
                descInputSubscription.unsubscribe();
            }
        });

    }

    private void buildPopupElement(ContextMenu menu, String s)
    {
        GridPane pane = new GridPane();

        ColumnConstraints c = new ColumnConstraints();
        c.setPercentWidth(100);
        pane.getColumnConstraints().add(c);
        Label innerButton = new Label(s);
        innerButton.setPadding(new Insets(5, 0, 0, 5));
        innerButton.setMaxWidth(Double.MAX_VALUE);
        innerButton.setMaxHeight(Double.MAX_VALUE);
        innerButton.setAlignment(Pos.CENTER_LEFT);

        pane.addRow(0, innerButton);
        pane.setPrefWidth(taskDescription.getWidth());
        pane.setPrefHeight(taskDescription.getHeight() * 0.75);
        CustomMenuItem customMenuItem = new CustomMenuItem(pane);
        menu.getItems().add(customMenuItem);
    }

    // build the UI
    private Node init()
    {
        Node node = loadFromFXML();
        connectActions(node, controller);
        connectMessageSource(node);
        return node;
    }

    private void toStartButton()
    {
        startButton.getStyleClass().setAll("btn", "btn-success");
    }

    private void toStopButton()
    {
        startButton.getStyleClass().setAll("btn", "btn-danger");
    }

    private void onFocus(TextField focusedField)
    {
        ContextMenu menu = new ContextMenu();
        menu.maxWidthProperty().bind(focusedField.widthProperty());
        menu.setAutoFix(true);

        EventStreams.changesOf(focusedField.focusedProperty()).subscribe(booleanChange ->
        {
            if (booleanChange.getNewValue())
            {
                menu.show(focusedField, Side.BOTTOM, 0, 0);
            }
            else
            {
                menu.hide();
            }
        });

        EventStreams.valuesOf(tasksToShow).subscribe(tasks ->
        {
            if (tasks == null)
            {
                return;
            }
            menu.getItems().clear();
            tasks.forEach(s -> buildPopupElement(menu, s));
            menu.show(focusedField, Side.BOTTOM, 0, 0);
        });
    }

}

File: griffon-app/views/ru/dkolmogortsev/ControlPanelView.kt
package ru.dkolmogortsev

import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIcon
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIconView
import griffon.core.artifact.GriffonView
import griffon.inject.MVCMember
import griffon.metadata.ArtifactProviderFor
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleListProperty
import javafx.collections.ObservableList
import javafx.fxml.FXML
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.geometry.Side
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.layout.ColumnConstraints
import javafx.scene.layout.GridPane
import org.codehaus.griffon.runtime.javafx.artifact.AbstractJavaFXGriffonView
import org.reactfx.EventStreams
import org.reactfx.Subscription
import ru.dkolmogortsev.messages.StartTask
import ru.dkolmogortsev.task.Task
import ru.dkolmogortsev.task.search.SearchFields
import ru.dkolmogortsev.task.search.SearchFields.TASKNAME

@ArtifactProviderFor(GriffonView::class)
class ControlPanelView : AbstractJavaFXGriffonView() {
    @FXML
    private lateinit var startButton: Button
    @FXML
    private lateinit var timerLabel: Label
    @FXML
    private lateinit var taskDescription: TextField
    @FXML
    private lateinit var taskName: TextField
    @FXML
    private lateinit var menuButton: Button
    @MVCMember
    private lateinit var controller: ControlPanelController
    @MVCMember
    private lateinit var model: ControlPanelModel
    @MVCMember
    private lateinit var parentView: ControlAndTaskView
    private val started = SimpleBooleanProperty(false)
    private val tasksToShow = SimpleListProperty<Task>()
    private val menu = ContextMenu()
    private lateinit var eventSub: Subscription
    private lateinit var popupSubscription: Subscription

    override fun initUI() {
        val node = init()
        timerLabel.textProperty().bind(model.timerTextProp)
        started.bind(model.taskStartedProperty())
        tasksToShow.bind(model.tasksProperty())

        taskDescription.textProperty().bindBidirectional(model.taskDescriptionProperty())
        taskName.textProperty().bindBidirectional(model.taskNameProperty())

        startButton.styleClass.clear()
        menuButton.styleClass.clear()
        val menuIcon = MaterialDesignIconView(MaterialDesignIcon.MENU)
        menuButton.heightProperty().addListener { _, _, newValue -> menuIcon.glyphSize = newValue.toDouble() / 2 }
        menuButton.graphic = menuIcon


        toStartButton()
        parentView.getContainerPane().addRow(0, node)
        started.addListener { _, _, newValue ->
            if (newValue) {
                toStopButton()
            } else {
                toStartButton()
            }
        }
        node.requestFocus()
        onFocus(taskDescription, SearchFields.DESCRIPTION)
        onFocus(taskName, TASKNAME)

    }

    private fun buildPopupElement(menu: ContextMenu, s: Task, mapper: (Task) -> String) {
        val pane = GridPane()

        val c = ColumnConstraints()
        c.percentWidth = 100.0
        pane.columnConstraints.add(c)
        val innerButton = Label(mapper.invoke(s))
        innerButton.padding = Insets(5.0, 0.0, 0.0, 5.0)
        innerButton.maxWidth = java.lang.Double.MAX_VALUE
        innerButton.maxHeight = java.lang.Double.MAX_VALUE
        innerButton.alignment = Pos.CENTER_LEFT

        pane.addRow(0, innerButton)
        pane.prefWidth = taskDescription.width
        pane.prefHeight = taskDescription.height * 0.75
        val customMenuItem = CustomMenuItem(pane)

        customMenuItem.setOnAction { event -> application.eventRouter.publishEvent(StartTask(s.id)) }

        menu.items.add(customMenuItem)
    }

    // build the UI
    private fun init(): Node {
        var node = loadFromFXML()!!
        connectActions(node, controller)
        connectMessageSource(node)
        return node
    }

    private fun toStartButton() {
        startButton.styleClass.setAll("btn", "btn-success")
        startButton.text = "Start"
    }

    private fun toStopButton() {
        startButton.styleClass.setAll("btn", "btn-danger")
        startButton.text = "Stop"
    }

    private fun onFocus(focusedField: TextField, searchField: SearchFields) {
        menu.maxWidthProperty().bind(focusedField.widthProperty())
        menu.isAutoFix = true

        EventStreams.changesOf(focusedField.focusedProperty()).subscribe { booleanChange ->
            if (booleanChange.newValue) {
                menu.show(focusedField, Side.BOTTOM, 0.0, 0.0)
                eventSub = EventStreams.valuesOf(focusedField.textProperty())
                        .subscribe { s -> controller.search(s, searchField) }
                if (focusedField.text == searchField.toString()) {
                    focusedField.text = ""
                }


                popupSubscription = EventStreams.valuesOf<ObservableList<Task>>(tasksToShow).subscribe { tasks ->
                    if (tasks == null) {
                    }
                    menu.items.clear()
                    tasks.stream().limit(10).forEach { s -> buildPopupElement(menu, s, searchField.mapper) }
                    if (tasks.size > 10) {
                        buildPopupElement(menu, Task("Found more than 10 elements", "Found more than 10 elements"), { task -> task.description })
                    }
                    menu.show(focusedField, Side.BOTTOM, 0.0, 0.0)
                }

            } else {
                popupSubscription.unsubscribe()
                eventSub.unsubscribe()
                menu.hide()
            }
        }


    }

}
